src/parsers/trades/gate.io/index.ts

```diff
@@ -0,0 +1,93 @@
+    let filledTrade = data[0];
+    let pointTrade = data[0];
+    let lineContinuity = 0;
+    for (const trade of data) {
+        const tradeToAdd: IPartialTrade = {};
+        tradeToAdd.date = createDateAsUTC(new Date(trade.Time)).getTime();
+        tradeToAdd.exchange = EXCHANGES.GateIO;
+        switch (trade['Action type']) {
+            case 'Trading Fees': {
+                lineContinuity = 1;
+                feeTrade = trade;
+                continue;
+            }
+            case 'Order Filled': {
+                if (lineContinuity++ !== 1) {
+                    console.error('Error parsing Gate.io trade lineContinuity++ !== 1 lineContinuity=${lineContinuity}');
+                    lineContinuity = 0;
+                    break;
+                }
+                filledTrade = trade;
+                continue;
+            }
+            case 'Order Placed': {
+                if (lineContinuity !== 2) {
+                    console.error('Error parsing Gate.io trade lineContinuity !== 2 lineContinuity=${lineContinuity}');
+                    lineContinuity = 0;
+                    break;
+                }
+                lineContinuity = 0;
+                tradeToAdd.boughtCurrency = filledTrade.Currency;
+                tradeToAdd.soldCurrency = trade.Currency;
+                tradeToAdd.amountSold = Math.abs(parseFloat(trade['Change amount']));
+                tradeToAdd.rate = Math.abs(parseFloat(trade['Change amount']) / parseFloat(filledTrade['Change amount']));
+                tradeToAdd.transactionFeeCurrency = feeTrade.Currency;
+                tradeToAdd.transactionFee = Math.abs(parseFloat(feeTrade['Change amount']));
+                break;
+            }
+            case 'Points Purchase': {
+                switch (lineContinuity) {
+                    case 0: {
+                        pointTrade = trade;
+                        lineContinuity = 1;
+                        continue;
+                    }
+                    case 1: {
+                        lineContinuity = 0;
+                        tradeToAdd.boughtCurrency = pointTrade.Currency;
+                        tradeToAdd.soldCurrency = trade.Currency;
+                        tradeToAdd.amountSold = Math.abs(parseFloat(trade['Change amount']));
+                        tradeToAdd.rate = Math.abs(parseFloat(trade['Change amount']) / parseFloat(pointTrade['Change amount']));
+                        break;
+                    }
+                    default: {
+                        console.error(`Error parsing Gate.io trade lineContinuity=${lineContinuity}`);
+                        break;
+                    }
+                }
+                break;
+            }
+            default: {
+                console.log(`Ignored Gate.io trade of type ${trade['Action type']}`);
+                continue;
+            }
+        }
+        tradeToAdd.ID = createID(tradeToAdd);
+        tradeToAdd.exchangeID = tradeToAdd.ID;
```

Comment:
I believe you should be able to move this up to the top. Also ideally we should set this before we create the ID

Reply:
So ID would be set to createID after the initial partial trade setting?
I modified the code from the binance parser and followed that structure.
How come exchangeID = trade.ID?

Reply:
Yes, and good point the binance parser should be updated. Its outdated and now not making IDs properly.

Ideally when we fix it we should bump the version and make part of the migration process to recompute all trade ids as they will all change then

Reply:
@starsoccer Please if you can tell your opinion about this:
"I also plan to update the Binance parser. I have exported the data from Binance with the "all statements" feature.
It contains operations of all different kinds.
In that case, should we import trades, transactions, and income all at once?
Or should I make a parser for each type, all of them taking only the concerned type of operations?"
It is not the case only with Binance, also with other platforms.

I would say import all at once, but I’m also ok with implementing it as separate, since it is the current structure.

[See all part.csv](https://github.com/starsoccer/cryptotithe/files/10220352/all.part.csv)

Reply:
I cant seem to open that file. Is that a zip or csv?

And regarding the binance question, I am fine with it all being in one parser, but we may need to do some tweaking to the current import design because I believe it asks for type first(transaction, income, or trade)

Reply:
Try this one (it’s the Binance one)

[all part.csv](https://github.com/starsoccer/cryptotithe/files/10251821/all.part.csv)

I will write it as an all-in-one importer then